src/meta-srv/src/region/supervisor.rs

```diff
@@ -0,0 +1,523 @@
+use crate::procedure::region_migration::manager::RegionMigrationManagerRef;
+use crate::procedure::region_migration::RegionMigrationProcedureTask;
+use crate::selector::SelectorOptions;
+
+/// `DatanodeHeartbeat` represents the heartbeat signal sent from a datanode.
+/// It includes identifiers for the cluster and datanode, a list of regions being monitored,
+/// and a timestamp indicating when the heartbeat was sent.
+#[derive(Debug)]
+pub(crate) struct DatanodeHeartbeat {
+    cluster_id: ClusterId,
+    datanode_id: DatanodeId,
+    // TODO(weny): Considers collecting the memtable size in regions.
+    regions: Vec<RegionId>,
+    timestamp: i64,
+}
+
+impl From<&Stat> for DatanodeHeartbeat {
+    fn from(value: &Stat) -> Self {
+        DatanodeHeartbeat {
+            cluster_id: value.cluster_id,
+            datanode_id: value.id,
+            regions: value.region_stats.iter().map(|x| x.id).collect(),
+            timestamp: value.timestamp_millis,
+        }
+    }
+}
+
+pub(crate) enum Event {
+    Tick,
+    HeartbeatArrived(DatanodeHeartbeat),
+    Clear,
+    #[cfg(test)]
+    Dump(tokio::sync::oneshot::Sender<RegionFailureDetector>),
+}
+
+impl Debug for Event {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Tick => write!(f, "Tick"),
+            Self::HeartbeatArrived(arg0) => f.debug_tuple("HeartbeatArrived").field(arg0).finish(),
+            Self::Clear => write!(f, "Clear"),
+            #[cfg(test)]
+            Self::Dump(_) => f.debug_struct("Dump").finish(),
+        }
+    }
+}
+
+pub type RegionSupervisorTickerRef = Arc<RegionSupervisorTicker>;
+
+/// A background job to generate [`Event::Tick`] type events.
+#[derive(Debug)]
+pub struct RegionSupervisorTicker {
+    /// The [`Option`] wrapper allows us to abort the job while dropping the [`RegionSupervisor`].
+    tick_handle: Mutex<Option<JoinHandle<()>>>,
+
+    /// The interval of tick.
+    tick_interval: Duration,
+
+    /// Sends [Event]s.
+    sender: Sender<Event>,
+}
+
+impl RegionSupervisorTicker {
+    /// Starts the ticker.
+    pub fn start(&self) {
+        let mut handle = self.tick_handle.lock().unwrap();
+        if handle.is_none() {
+            let sender = self.sender.clone();
+            let tick_interval = self.tick_interval;
+            let ticker_loop = tokio::spawn(async move {
+                let mut interval = interval(tick_interval);
+                interval.set_missed_tick_behavior(MissedTickBehavior::Skip);
+                if let Err(err) = sender.send(Event::Clear).await {
+                    warn!(err; "EventReceiver is dropped, failed to send Event::Clear");
+                    return;
+                }
+                loop {
+                    interval.tick().await;
+                    if let Err(err) = sender.send(Event::Tick).await {
+                        warn!(err; "EventReceiver is dropped, tick loop is stopped");
```

Comment:
Looks like it should be `info!` instead of `warn!`?

Reply:
Yes